### PR TITLE
Kotlin Interview Test

### PR DESCRIPTION
**Pull request by gooberholtzer**
Finally had some time to "properly" look over Kotlin.  Let me know what I still have effed up.   

builds with circleci

### CODE REVIEW

File: src/main/kotlin/io/hammerhead/interview/App.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*
+ * This Kotlin source file was generated by the Gradle 'init' task.
+ */
+package io.hammerhead.interview
+
+import io.hammerhead.interview.service.*
+
+class App {
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/App.kt]: What is this for? Probably just needs cleaned up.

File: src/main/kotlin/io/hammerhead/interview/RouteService.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package io.hammerhead.interview
+
+
+abstract class RouteService( val name : String, val userId : String, val routes : List<String> ) {
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/RouteService.kt]: Organizationally in seems like creating a route service with the `userId` prevents an instance of this class from being re-used.

File: src/main/kotlin/io/hammerhead/interview/RouteService.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    fun getAllRoutes() : List<String> = routes
+
+    abstract fun getUserRoutes() : List<String>
+
+    fun getRoutes( adornRoutes : (routeName: String) -> String) : List<String> {
+
+        val userRoutes = mutableListOf<String>()
+        for( route in getAllRoutes()) {
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/RouteService.kt]: You should look into the `map` function for a more concise way of writing this modification of a list.

File: src/main/kotlin/io/hammerhead/interview/service/RouteServiceAggregator.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package io.hammerhead.interview.service
+
+import io.hammerhead.interview.RouteService
+
+class RouteServiceAggregator( val routeServices: List<RouteService>) {
+
+    fun aggregateItems(routeTreatment : (_routeService: RouteService) -> List<String>) : List<String> {
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/service/RouteServiceAggregator.kt]: Seems like this is only used by this class and is more of a helper method so it should be `private` and at the end.

Comment (brentnd) [reply src/main/kotlin/io/hammerhead/interview/service/RouteServiceAggregator.kt]: Also it is another case that simplifies to a `map` call.

File: src/test/kotlin/io/hammerhead/interview/AppTest.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/*
+ * This Kotlin source file was generated by the Gradle 'init' task.
+ */
+package io.hammerhead.interview
+
+import kotlin.test.Test
+import kotlin.test.assertNotNull
+
+class AppTest {
```
Comment (brentnd) [inline src/test/kotlin/io/hammerhead/interview/AppTest.kt]: Boilerplate code probably not worth including.

File: src/main/kotlin/io/hammerhead/interview/service/RouteServiceAggregator.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        val allItems = mutableListOf<String>()
+
+        for( routeService in routeServices)
+            allItems +=  routeTreatment(routeService)
+
+        return allItems
+    }
+
+    fun getAllRoutes() : List<String> = aggregateItems({ it.routes })
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/service/RouteServiceAggregator.kt]: With Kotlin, when you implement a function with an expression body the return type can be inferred.

File: src/main/kotlin/io/hammerhead/interview/App.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    var greeting = "I'm here, though this is not part of the interview."
+}
+
+fun main( args: Array<String>) {
+
+    val userId = "42"
+
+    // passing in the userId to each service ot leave room for that
+    // to be different per-service.
+    var routeServiceAgg = RouteServiceAggregator(listOf( Strava(userId), Komoot(userId), RWGPS(userId)))
```
Comment (brentnd) [inline src/main/kotlin/io/hammerhead/interview/App.kt]: Here is the easiest place to see that passing in the `userId` at this point would make it impossible to call without reconstructing the aggregator with different service instances for another user.

Comment (gooberholtzer) [reply src/main/kotlin/io/hammerhead/interview/App.kt]: understood.  was kind of going for "immutability" but, again. I understand.

**Comment by brentnd** [comment]
@gooberholtzer thanks for the submission.

Closing this and we will follow up with you.

**Pull request closed without merging**